### DISCUSSION TITLE
`useAtom(undefined)` throws `Invalid value used as weak map key`

### DISCUSSION THREAD

**Discussion by kevinschaich** [pmndrs/jotai#2777] [Ideas]
This should be checked for or caught and re-thrown as a more helpful error message.

To Repro:

```tsx
const value = useAtom(undefined)

    -> TypeError: Invalid value used as weak map key
```

I hit this today and if it weren't for this comment I would have been very struggling for a few hours:
https://github.com/pmndrs/jotai/discussions/1816#discussioncomment-9600796

One potentially common way people might hit this is if they use nested/split atoms and try to reference a key that doesn't exist.

e.g. objects:
```tsx
const objectAtom = atom({
    a: atom(...)
    b: atom(...)
)

...

const bValue = useAtomValue(useAtomValue(objectAtom)[c])
```

e.g. lists:
```tsx
const listAtom = atom([atom(...), atom(...)]

...

const thirdValue = useAtomValue(useAtomValue(listAtom)[2])
```

Related: https://github.com/pmndrs/jotai/issues/2272, https://github.com/pmndrs/jotai/issues/1817, https://github.com/pmndrs/jotai/issues/1177, https://github.com/pmndrs/jotai/discussions/1816, https://github.com/pmndrs/jotai/discussions/2239, https://github.com/pmndrs/jotai/discussions/2273

**Comment by kevinschaich** [comment]
cc: @dai-shi I believe you replied to a few of the issues/discussions above.

**Reply by dai-shi** [reply]
Sure. Would you like to open a PR? Or just suggesting an improved error message may help.

I also have a question for you. Are you not using TS in this case? Because TS can warn you, we haven't prioritize this behavior much.

**Reply by kevinschaich** [reply]
I am using TypeScript but it is not bulletproof. If your signature for a dynamic object is

```tsx
{ [key: string]: Atom<...> }
```

you could pass in any string key and Typescript will not complain until it fails at useAtom.

**Reply by dai-shi** [reply]
yeah, unless you enable `noUncheckedIndexedAccess`.

**Reply by kevinschaich** [reply]
This is disabled as the default and marked as not planned (https://github.com/microsoft/TypeScript/issues/49169).

**Comment by kevinschaich** [comment]
Fixed in https://github.com/pmndrs/jotai/pull/2778
